cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -80,6 +80,160 @@ const FunctionDoc list_value_length_doc{
+  AddListElementArrayKernel<ListType, Int16Type>(func);
+  AddListElementArrayKernel<ListType, UInt32Type>(func);
+  AddListElementArrayKernel<ListType, Int32Type>(func);
+  AddListElementArrayKernel<ListType, UInt64Type>(func);
+  AddListElementArrayKernel<ListType, Int64Type>(func);
+
+  AddListElementArrayKernel<LargeListType, UInt8Type>(func);
+  AddListElementArrayKernel<LargeListType, Int8Type>(func);
+  AddListElementArrayKernel<LargeListType, UInt16Type>(func);
+  AddListElementArrayKernel<LargeListType, Int16Type>(func);
+  AddListElementArrayKernel<LargeListType, UInt32Type>(func);
+  AddListElementArrayKernel<LargeListType, Int32Type>(func);
+  AddListElementArrayKernel<LargeListType, UInt64Type>(func);
+  AddListElementArrayKernel<LargeListType, Int64Type>(func);
+
+  AddListElementArrayKernel<FixedSizeListType, UInt8Type>(func);
+  AddListElementArrayKernel<FixedSizeListType, Int8Type>(func);
+  AddListElementArrayKernel<FixedSizeListType, UInt16Type>(func);
+  AddListElementArrayKernel<FixedSizeListType, Int16Type>(func);
+  AddListElementArrayKernel<FixedSizeListType, UInt32Type>(func);
+  AddListElementArrayKernel<FixedSizeListType, Int32Type>(func);
+  AddListElementArrayKernel<FixedSizeListType, UInt64Type>(func);
+  AddListElementArrayKernel<FixedSizeListType, Int64Type>(func);
+}
+
+void AddListElementScalarKernels(ScalarFunction* func) {
+  AddListElementScalarKernel<ListScalar, UInt8Type>(func);
+  AddListElementScalarKernel<ListScalar, Int8Type>(func);
+  AddListElementScalarKernel<ListScalar, UInt16Type>(func);
+  AddListElementScalarKernel<ListScalar, Int16Type>(func);
+  AddListElementScalarKernel<ListScalar, UInt32Type>(func);
+  AddListElementScalarKernel<ListScalar, Int32Type>(func);
+  AddListElementScalarKernel<ListScalar, UInt64Type>(func);
+  AddListElementScalarKernel<ListScalar, Int64Type>(func);
+
+  AddListElementScalarKernel<LargeListScalar, UInt8Type>(func);
+  AddListElementScalarKernel<LargeListScalar, Int8Type>(func);
+  AddListElementScalarKernel<LargeListScalar, UInt16Type>(func);
+  AddListElementScalarKernel<LargeListScalar, Int16Type>(func);
+  AddListElementScalarKernel<LargeListScalar, UInt32Type>(func);
+  AddListElementScalarKernel<LargeListScalar, Int32Type>(func);
+  AddListElementScalarKernel<LargeListScalar, UInt64Type>(func);
+  AddListElementScalarKernel<LargeListScalar, Int64Type>(func);
+
+  AddListElementScalarKernel<FixedSizeListScalar, UInt8Type>(func);
+  AddListElementScalarKernel<FixedSizeListScalar, Int8Type>(func);
+  AddListElementScalarKernel<FixedSizeListScalar, UInt16Type>(func);
+  AddListElementScalarKernel<FixedSizeListScalar, Int16Type>(func);
+  AddListElementScalarKernel<FixedSizeListScalar, UInt32Type>(func);
+  AddListElementScalarKernel<FixedSizeListScalar, Int32Type>(func);
+  AddListElementScalarKernel<FixedSizeListScalar, UInt64Type>(func);
+  AddListElementScalarKernel<FixedSizeListScalar, Int64Type>(func);
```

Comment:
I need the list type to define the scalar type for the kernel signature:
```cpp
template <typename InListScalarType, typename IndexType>
void AddListElementScalarKernel(ScalarFunction* func) {
  auto inputs = {InputType::Scalar(InListScalarType::TypeClass::type_id),
                 InputType::Scalar(IndexType::type_id)};
...
```

Reply:
I think we can do something like this then:

```cpp
void AddListElementScalarKernels(ScalarFunction* func) {
  for (const auto list_type_id : {Type::LIST, Type::LARGE_LIST, Type::FIXED_SIZE_LIST}) {
    for (const auto& index_type : IntTypes()) {
    auto inputs = {InputType::Scalar(list_type_id), InputType::Scalar(index_type)};
    auto output = OutputType{ListValuesType};
    auto sig =
        KernelSignature::Make(std::move(inputs), std::move(output), /*is_varargs=*/false);
    auto scalar_exec = GenerateInteger<ListElementScalar, void>(index_type);
    ScalarKernel kernel{std::move(sig), std::move(scalar_exec)};
    kernel.null_handling = NullHandling::COMPUTED_NO_PREALLOCATE;
    kernel.mem_allocation = MemAllocation::NO_PREALLOCATE;
    DCHECK_OK(func->AddKernel(std::move(kernel)));
  }
  }
}
```

This will register all the kernels in one go and save us some templating, and is more in line with what we do when we generate a kernel for many different types.

GenerateInteger comes from codegen_internal.h: https://github.com/apache/arrow/blob/e5f3e04b4b80c9b9c53f1f0f71f39d9f8308dced/cpp/src/arrow/compute/kernels/codegen_internal.h#L1018-L1022

IntTypes comes from type.h: https://github.com/apache/arrow/blob/e5f3e04b4b80c9b9c53f1f0f71f39d9f8308dced/cpp/src/arrow/type.h#L1966-L1967

The kernel itself would look something like this:

```cpp
template <typename, typename IndexType>
Status ListElementScalar(KernelContext* /*ctx*/, const ExecBatch& batch, Datum* out) {
  using IndexScalarType = typename TypeTraits<IndexType>::ScalarType;
  const auto& index_scalar = batch[1].scalar_as<IndexScalarType>();
  if (ARROW_PREDICT_FALSE(!index_scalar.is_valid)) {
    return Status::Invalid("Index must not be null");
  }
  // note the below - no need to cast to the concrete scalar type, the base class is enough
  const auto& list_scalar = batch[0].scalar_as<BaseListScalar>();
  // ...snip...
  return Status::OK();
}
```

Reply:
This is good stuff!
Out of curiosity: Do you know what other kernels have this structure? or what other kernels need to cover all types for one of its inputs?

Reply:
From a quick search, how about cast kernels? Here a cast from any integer to Decimal128 is implemented via GenerateInteger: https://github.com/apache/arrow/blob/293f85641643b170a3557d6d99eaab8f18109a6a/cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc#L676-L692

Reply:
@lidavidm Are you sure that compiles? It seems the kernel is invalid as generator template

Reply:
Now it compiles, I needed a wrapper to have a well defined generator
```cpp
template <typename, typename IndexType>
struct ListElementScalar{
  static Status Exec
```